File: CamStudy/CamStudy/ViewModels/GoalMatesViewModel.swift
import Foundation

@MainActor
final class GoalMatesViewModel: ObservableObject {
    @Published var mates: [UserProfile] = []
    @Published var errorMessage: String?
    @Published var isLoading = false

    private let firestoreService = FirestoreService.shared

    func loadMates(goal: String, excludingUserId: String) {
        isLoading = true
        errorMessage = nil
        firestoreService.fetchGoalMates(goal: goal, excludingUserId: excludingUserId) { [weak self] result in
            Task { @MainActor in
                guard let self else { return }
                self.isLoading = false
                switch result {
                case .success(let mates):
                    self.mates = mates
                case .failure(let error):
                    self.errorMessage = error.localizedDescription
                }
            }
        }
    }
}

File: camstudy_app/lib/viewmodels/goal_mates_view_model.dart
import 'package:flutter/foundation.dart';

import '../models/user_profile.dart';
import '../services/firestore_service.dart';

class GoalMatesViewModel extends ChangeNotifier {
  List<UserProfile> mates = [];
  String? errorMessage;
  bool isLoading = false;

  final _firestoreService = FirestoreService.shared;

  Future<void> loadMates({
    required String goal,
    required String excludingUserId,
  }) async {
    isLoading = true;
    errorMessage = null;
    notifyListeners();

    try {
      mates = await _firestoreService.fetchGoalMates(
        goal: goal,
        excludingUserId: excludingUserId,
      );
    } catch (e) {
      errorMessage = e.toString();
    }
    isLoading = false;
    notifyListeners();
  }
}
